packages/@aws-cdk/aws-iam/lib/policy-statement.ts

```diff
@@ -821,4 +827,51 @@ export function deriveEstimateSizeOptions(scope: IConstruct): EstimateSizeOption
   }
 
   return { actionEstimate, arnEstimate };
+}
+
+/**
+ * A class that behaves both as a set and an array
+ *
+ * Used for the elements of a PolicyStatement. In practice they behave as sets,
+ * but we have thousands of snapshot tests in existence that will rely on a
+ * particular order so we can't just change the type to `Set<>` wholesale without
+ * causing a lot of churn.
+ */
+class OrderedSet<A> {
+  private readonly set = new Set<A>();
+  private readonly array = new Array<A>();
+
+  /**
+   * Add new elements to the set
+   *
+   * Returns the elements actually added
+   */
+  public push(...xs: A[]) {
+    const ret = new Array<A>();
+    for (const x of xs) {
+      if (!this.set.has(x)) {
+        this.set.add(x);
+        this.array.push(x);
+        ret.push(x);
+      }
```

Comment:
Personal taste... but I prefer short-cutting the other way around (not that it changes anything much):

```suggestion
      if (this.set.has(x)) {
        continue;
      }
      this.set.add(x);
      this.array.push(x);
      ret.push(x);
```